packages/adminrouter/extra/src/test-harness/tests/test_service.py

```diff
@@ -0,0 +1,473 @@
+# Copyright (C) Mesosphere, Inc. See LICENSE file for details.
+
+import time
+
+import pytest
+import requests
+
+from generic_test_code.common import (
+    generic_correct_upstream_dest_test,
+    generic_correct_upstream_request_test,
+    header_is_absent,
+)
+from mocker.endpoints.marathon import NGINX_APP_ALWAYSTHERE_DIFFERENTPORT
+from mocker.endpoints.mesos import (
+    NGINX_FWRK_ALWAYSTHERE_DIFFERENTPORT,
+    NGINX_FWRK_ALWAYSTHERE_NOWEBUI,
+    framework_from_template,
+)
+from mocker.endpoints.mesos_dns import NGINX_SRV_ALWAYSTHERE_DIFFERENTPORT
+
+CACHE_UPDATE_DELAY = 2  # seconds
+assert CACHE_UPDATE_DELAY > 1.5  # due to cache_expiration=(CACHE_UPDATE_DELAY - 1)
+
+
+@pytest.fixture()
+def master_ar_process_fastcache(nginx_class):
+    """An AR process instance fixture for situations where a short lived cache
+       is needed. In order to achieve it, the scope of this fixture needs to be
+       single test.
+    """
+    nginx = nginx_class(role="master",
+                        cache_poll_period=CACHE_UPDATE_DELAY,
+                        cache_expiration=(CACHE_UPDATE_DELAY - 1),
+                        cache_first_poll_delay=1,
+                        )
+    nginx.start()
+
+    yield nginx
+
+    nginx.stop()
+
+
+class TestServiceStatefull:
+    # Test all the statefull test-cases/tests where AR caching may influence the
+    # results
+    def test_if_marathon_apps_are_resolved(
+            self, master_ar_process_fastcache, mocker, valid_user_header):
+        # Remove the data from MesosDNS and Mesos mocks w.r.t. resolved service
+        mocker.send_command(endpoint_id='http://127.0.0.2:5050',
+                            func_name='set_frameworks_response',
+                            aux_data=[])
+        mocker.send_command(endpoint_id='http://127.0.0.1:8123',
+                            func_name='set_srv_response',
+                            aux_data={})
+
+        # Check if we can use root Marathon app data to resolve `/service`
+        # location requests:
+        generic_correct_upstream_dest_test(
+            master_ar_process_fastcache,
+            valid_user_header,
+            '/service/nginx-alwaysthere/foo/bar/',
+            "http://127.0.0.1:16000"
+            )
+
+        # Update the application, wait for cache to register the change
+        new_apps = {"apps": [NGINX_APP_ALWAYSTHERE_DIFFERENTPORT, ]}
+        mocker.send_command(endpoint_id='http://127.0.0.1:8080',
+                            func_name='set_apps_response',
+                            aux_data=new_apps)
+        time.sleep(CACHE_UPDATE_DELAY + 1)
+
+        # Check if the location now resolves correctly to the new app port
+        generic_correct_upstream_dest_test(
+            master_ar_process_fastcache,
+            valid_user_header,
+            '/service/nginx-alwaysthere/foo/bar/',
+            "http://127.0.0.15:16001"
+            )
+
+    def test_if_webui_url_is_resolved_using_framework_id(
```

Comment:
`webui_url` is a field present in `/mesos/state-summary` endpoint output. It can be used for routing requests, though Marathon labels have priority.